flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -596,6 +596,7 @@ && isChainableInput(sourceOutEdge, streamGraph)) {
                                                     chainedSources,
                                                     streamGraph));
                     chainInfo.addCoordinatorProvider(coord);
+                    chainInfo.getAllChainedNodes().add(streamGraph.getStreamNode(sourceNodeId));
```

Comment:
Modifying a collection returned by a getter is not recommended.

I would propose to introduce a method `StreamNode OperatorChainInfo#recordChainedNode(int nodeId)` and reuse it in `OperatorChainInfo#addNodeToChain(int, String)`.